### ISSUE TITLE
Default output type of vcf2phylip.py: too many ambiguous nucleotide sequences?

### ISSUE THREAD

**Issue by LiuCanidk** [edgardomortiz/vcf2phylip#51]
Hi, thanks for developing this tool

I run the script of vcf2phylip.py successfully but found the output seems to be the amino acid sequences. My code and the screenshort of my output file are as follows:

> python /work/share/acuwbf4fll/liucan/software/phylip/vcf2phylip-2.8/vcf2phylip.py -i /work/share/acuwbf4fll/liucan/HND_project/Bulk_RNA_variant_calling/06.GVCF_filter/output/HND.SNV.recode.vcf --output-folder /work/share/acuwbf4fll/liucan/HND_project/Bulk_RNA_variant_calling/09.phylotree --output-prefix HND_RNA_SNV

![image](https://github.com/user-attachments/assets/6429d8cb-18a3-4f42-9121-96230870f0dc)

I did not found any parameters specified to set the output type, but I prefer the nucelotide sequences alignment to be output. How can I do for this?

Any suggestions would be greatly appreciated!

**Comment by edgardomortiz** [comment]
They are nucleotides, VCF doesn't support aminoacids as far as I know. Your heterozygous genotypes are represented with ambiguity codes, see here https://www.promega.com/resources/guides/nucleic-acid-analysis/restriction-enzyme-resource/restriction-enzyme-resource-tables/iupac-ambiguity-codes-for-nucleotide-degeneracy/

Edgardo

**Comment by LiuCanidk** [comment]
> They are nucleotides, VCF doesn't support aminoacids as far as I know. Your heterozygous genotypes are represented with ambiguity codes, see here https://www.promega.com/resources/guides/nucleic-acid-analysis/restriction-enzyme-resource/restriction-enzyme-resource-tables/iupac-ambiguity-codes-for-nucleotide-degeneracy/
> 
> Edgardo

@edgardomortiz  I see. Thanks for your reply! I then wonder if it is normal that my translated phylip file was filled with ambigous code and whether this wound affect the process of tree construction. If so, then should I enable the parameter of `--resolve-IUPAC` to choose one nucleotide forcely？

**Comment by edgardomortiz** [comment]
I don't think it is a good idea to translate SNPs, they are not contiguous in the genome. Besides that, degenerate nucleotides will create degenerate aminoacids as well during translation. The option --resolve-IUPAC will choose one nucleotide at random when you have an ambiguity, you may try that but I think I won't fix your issue of trying to translate SNPs (unless I am missing something about your specific VCF).

I hope this makes sense,

Edgardo

**Comment by LiuCanidk** [comment]
> I don't think it is a good idea to translate SNPs, they are not contiguous in the genome. Besides that, degenerate nucleotides will create degenerate aminoacids as well during translation. The option --resolve-IUPAC will choose one nucleotide at random when you have an ambiguity, you may try that but I think I won't fix your issue of trying to translate SNPs (unless I am missing something about your specific VCF).
> 
> I hope this makes sense,
> 
> Edgardo

Thanks for your reply. By stating "translating SNPs", I mean translating from the VCF format to a format of alignment, e.g., phylip format, for tree construction. There may be some misleading that I did not mean translating from nucleotides to amino acid sequences. Sorry about that.

I agree that SNPs are discontinuous in the genome. I am just wondering why I got so many ambiguous sequences from VCF format and whether I should add the `-resolve-IUPAC` parameter to avoid this situation. That is, would too many ambiguous sequences hamper the downstream analysis of tree construction?

Thanks in advance

**Comment by edgardomortiz** [comment]
Ah I see, you meant _converting_ VCF to another format (sorry for being pedantic but translating has a biological meaning and I got confused). As I said above, you have heterozygous genotypes because I assume your organism is at least diploid. For phylogenetics it is common to use a single sequence per sample, the way to achieve this is by representing both possible nucleotides with a single ambiguity code. As for the consequences of these ambiguities on your data I can't predict them because I am obviously not familiar with the organisms you are analyzing, but in general I could say the more ambiguities the less resolved a tree ends up.

Maybe your SNP calling settings were set up incorrectly? Maybe your reference genome is too distant? I don't know, I am just speculating here...

Edgardo

**Comment by LiuCanidk** [comment]
Oh, sorry about the information loss. The organism is human, and more specificly, the material is a cancer cell line and of course with some treatments.

I checked the VCF file and did find something weird: some genotypes are missing, maybe it is the cause and may be due to hard genotype filtration.
![image](https://github.com/user-attachments/assets/eb07d0ba-25db-48b8-a979-2d9343904bed)

However, I wonder whether what you said about representing both possible nucleotides with a single ambiguity code could work. How can I achieve this?

**Comment by edgardomortiz** [comment]
> However, I wonder whether what you said about representing both possible nucleotides with a single ambiguity code could work. How can I achieve this?

This is what the script does by default, the reason you have the ambiguity codes in the first place. No need to do anything additional...

**Comment by cwz12** [comment]
> 你好，感谢你开发了这个工具
> 
> 我成功运行了 vcf2phylip.py 的脚本，但发现输出似乎是氨基酸序列。我的代码和输出文件的截屏如下：
> 
> > Python /work/share/acuwbf4fll/liucan/software/phylip/vcf2Phylip-2.8/vcf2phylip.py -I /work/share/acuwbf4fll/liucan/HND_project/Bulk_RNA_variant_calling/06.GVCF_filter/output/HND.SNV.recode.vcf --output-folder /work/share/acuwbf4fll/liucan/HND_project/Bulk_RNA_variant_calling/09.phylotree --output-prefix HND_RNA_SNV
> 
> ![image](https://github.com/user-attachments/assets/6429d8cb-18a3-4f42-9121-96230870f0dc)
> 
> 我没找到任何参数来设置输出类型，但我更倾向于输出核蛋白序列的比对。我该怎么做？
> 
> 任何建议都将不胜感激！

Have you managed to resolve this issue? When converting a VCF file containing filtered biallelic SNPs from whole-genome resequencing into FASTA format and constructing a maximum likelihood (ML) phylogenetic tree using IQ-TREE, I found that the resulting FASTA file contains many ambiguous IUPAC codes. As a result, IQ-TREE misidentifies the sequences as amino acid sequences. Although I forced IQ-TREE to treat the sequences as DNA using the -st DNA option, there are still problems with the tree construction.

<img width="1363" height="740" alt="Image" src="https://github.com/user-attachments/assets/5c15d401-74cb-4e21-a617-e0d41f703d59" />

**Comment by edgardomortiz** [comment]
Hi, there is no issue to fix here. The "problem" comes from having heterozygous (diploid or polyploid) SNP genotypes and collapsing them to a single sequence (i.e., representing more than one nucleotide as a single letter, the IUPAC ambiguity code). If you want to avoid IUPAC ambiguity codes you can add the option `--resolve-IUPAC` which will choose one of the nucleotides in the heterozygous call at random so only A C T G are written to the file, of course losing information in the process.

I hope this helps

Edgardo
